### ISSUE TITLE
Shuffle the train_dataset after Every Epoch

### ISSUE THREAD

**Issue by VarunGumma** [axolotl-ai-cloud/axolotl#2967]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Shuffling the training dataset after every epoch gives the best chance for the model to learn without any memorization or oscillating/stepwise loss pattern. Slightly surprising most of the training libraries don't provide any interface for this. 

### ✔️ Solution

A simple flag like `shuffle_train_dataset_every_epoch: bool | None` should do the trick. But I am unsure where exactly to add this so that all the trainers can make use of this. However, shuffling doesn't make sense with `group_by_length`.

### ❓ Alternatives

NA

### 📝 Additional Context

NA

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by winglian** [comment]
Hey @VarunGumma,we already shuffle via the random sampler on each epoch.
